torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py

```diff
@@ -162,38 +161,44 @@ def maybe_increase_iter(self, bucket):
 
 
 def powerSGD_hook(state: PowerSGDState, bucket) -> torch.futures.Future:
-    """
+    r"""
     This DDP communication hook implements the original PowerSGD gradient compression
     algorithm described in https://arxiv.org/abs/1905.13727.
     Once gradient tensors are aggregated across all workers, this hook applies
     compression as follows:
-    1) Views the input flattened 1D gradient tensor as two groups of per-parameter tensors:
-    high-rank tensors and vector-like rank-1 tensors (for biases).
-    2) Handles rank-1 tensors by allreducing them without compression:
-        2.1) Allocate contiguous memory for those rank-1 tensors,
-        and allreduces all the rank-1 tensors as a batch, without compression;
-        2.2) Copies the individual rank-1 tensors from the contiguous memory back to the input tensor.
-    3) Handles high-rank tensors by PowerSGD compression:
-        3.1) For each high-rank tensor M, creates two low-rank tensors P and Q for decomposing M,
+
+    1. Views the input flattened 1D gradient tensor as two groups of per-parameter tensors: high-rank tensors and vector-like rank-1 tensors (for biases).
+
+    2. Handles rank-1 tensors by allreducing them without compression:
+
+        2.1. Allocate contiguous memory for those rank-1 tensors, and allreduces all the rank-1 tensors as a batch, without compression;
+
+        2.2. Copies the individual rank-1 tensors from the contiguous memory back to the input tensor.
```

Comment:
Maybe point 1 and 2 could be summarized as simply:

```
1. Rank-1 gradients (corresponding to bias parameters) are not compressed and averaged between workers with regular all-reduce communication.
```

I feel like the allocation of contiguous memory is an implementation detail.

Reply:
Prefer keeping the details here. This was once asked by a code reviewer, as the pseudocode in the paper does not  expose such details.

There is actually another API `allreduce_coalesced` that can potentially allreduce these rank-1 tensors without explicit packing. Although this API is not supported on NCCL yet, the details here can remind that here can be a good use case.